File: exercises/recursive-factorial.js
/* 
TASK: implement a factorial function using recursion, memoization, closure

int -> int

example: memoizedFactorials(3) -> returns 6 
*/


const memoizedFactorials = memoizedFactorial();

console.log(`factorial of 1 is ${memoizedFactorials(1)}`);
console.log(`factorial of 4 is ${memoizedFactorials(4)}`);
console.log(`factorial of 100 is ${memoizedFactorials(100)}`);
console.log(`factorial of 1 is ${memoizedFactorials(1)}`);

//**************************
function memoizedFactorial() {
    let factorialCache = {};
    return checkCacheOrCompute;

    // *************************
    function checkCacheOrCompute(n) {
        if (n in factorialCache) {
            console.log(`fetching from factorialCache[${n}]:`)
            return factorialCache[n];
        } else {
            console.log(`computing...`)
            return computeFactorial(n);
        }
    }
    // *************************
    function computeFactorial(n) {
        let term = n
        if (term == 1) {
            factorialCache[n] = 1;
            return factorialCache[n];
        } else {
            factorialCache[n] = term * checkCacheOrCompute(n - 1);
            return factorialCache[n];
        }
    }
}

File: python exercises/recursive-factorial.py

def memoized_factorial(): 
    factorial_cache = {}

    def check_cache_or_compute(n): 
        if (n in factorial_cache):
            return factorial_cache[n]
        else:
            return compute_factorial(n)

    def compute_factorial(n):
        term = n
        if (term == 1):
            factorial_cache[n] = 1
            return factorial_cache[n]
        else:
            factorial_cache[n] = term * check_cache_or_compute(n - 1)
            return factorial_cache[n]

    return check_cache_or_compute


memoized_factorials = memoized_factorial()

print("The factorial of 1 is", memoized_factorials(1))
print("The factorial of 4 is", memoized_factorials(4))
print("The factorial of 7 is", memoized_factorials(7))
print("The factorial of 2 is", memoized_factorials(2))
print("The factorial of 1 is", memoized_factorials(1))